fetch git config correctly

dorianmariefr opened the pull request:
for instance user.name config was not fetched

git --config user.name was not a valid command

----

found this because i didn't have my name in the generated LICENSE file for instance

I built crystal from source after this change and it now works correctly

i didn't find any test that and i'm quite new to crystal so not sure how i would make a test

bcardiff commented:
Nice catch! That's why last time crystal init didn't work for me 😅

It seems to be introduced in #9141 where `git config --get #{key}` was used.

I'm not sure if there are any sensible differences between using or not --get in the case there are multiple values defined at local, global or system level.

Would you mind adding the `--get` to apply the changes #9141 should have?

bcardiff commented:
If you are wiling to write a spec the steps would be

1. create gitconfig file to be used in the spec (otherwise everybody will need to agree on the configuration)
2. Override the GIT_CONFIG env variable so git config subprocess will pick the config in (1). `spec/support/env.cr` offers a nice helper method to do that.
3. Add spec to `spec/compiler/crystal/tools/init_spec.cr`. The existing specs initialize the a `Crystal::Init::InitProject` with a `Crystal::Init::Config` instance already. It seems that the methods `Crystal::Init.run/parse_args/fetch_email` are not covered by the existing specs. Whatever coverage you are willing to add will be an improvement.

dorianmariefr commented:
thanks for the hints, i added a spec (thanks to oprypin, dscottboggs_gitlab, and @Blacksmoke16 for the help).

I also added email to the license but i can remove that

bcardiff commented on spec/compiler/crystal/tools/init_spec.cr:
```diff
@@ -6,6 +6,8 @@ require "ini"
 require "yaml"
 require "../../../support/tempfile"
+require "../../../support/env"
+require "../../../support/win32"
```
Was there any reason to include `support/win32` here?

dorianmariefr commented:
Yes otherwise I get an error for a missing method

On Wed 29 Jul 2020 at 14:33, Brian J. Cardiff <notifications@github.com>
wrote:

> *@bcardiff* commented on this pull request.
> ------------------------------
>
> In spec/compiler/crystal/tools/init_spec.cr
> <https://github.com/crystal-lang/crystal/pull/9640#discussion_r462262699>:
>
> > @@ -6,6 +6,8 @@ require "ini"
>  require "spec"
>  require "yaml"
>  require "../../../support/tempfile"
> +require "../../../support/env"
> +require "../../../support/win32"
>
> Was there any reason to include support/win32 here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/crystal-lang/crystal/pull/9640#pullrequestreview-457474472>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOASD5YU7XQCY5T3EVNLBLLR6AJJZANCNFSM4PK46ZIA>
> .
>
